rfcs/text/0013_saved_object_migrations.md

```diff
@@ -0,0 +1,511 @@
+they capture the current hr time `current_hr_time` and starts waiting until
+`process.hrtime() - current_hr_time > expiry_timeout` if at that point
+`lastSeen` hasn’t been updated the lease is considered to have expired. This
+means other nodes can take up to `2*expiry_timeout` to recognize an expired
+lease, but a lease will never expire prematurely. 
+
+Any node that detects an expired lease can release that lease by setting the
+expired node’s `hasLease = false`. It can then attempt to acquire its lease. 
+
+## 4.4 Minimize data loss with mixed Kibana versions
+When multiple versions of Kibana are running at the same time, writes from the
+outdated node can end up either in the outdated Kibana index, the newly
+migrated index, or both. New documents added (and some updates) into the old
+index while a migration is in-progress will be lost. Writes that end up in the
+new index will be in an outdated format. This could cause queries on the data
+to only return a subset of the results which leads to incorrect results or
+silent data loss.
+
+Any in-progress writes such as a bulk operation will continue to run even
+after an index was set to read-only or an alias had been removed. There are
+also certain failure scenarios under which ClusterState (such as aliases or
+index settings) remains inconsistent between nodes for an unbounded
+amount of time. Since Kibana cannot rely on Elasticsearch scripting support
+being enabled, we don't have a mechanism to guarantee that there would be no
+data loss. Instead, we aim to minimize it as much as possible.
+
+To minimize data loss we need to implement a “guard” which will prevent old
+nodes and new nodes from writing at the same time. To perform a rollback for a
+failed migration, this guard will have to be removed. By locating this state
+in the backup snapshot taken before an upgrade, rolling back to the previous
+snapshot will automatically remove the guard.
+
+**7.x Minimize data loss when creating new indices**
+
+Preventing data loss from mixed 7.x versions, introduces two additional steps
+to rollback to a previous version without a snapshot:
+1. (existing) Point the `.kibana` alias to the previous Kibana index `.kibana_n-1`
+2. (existing) Delete `.kibana_n`
+3. (new) Enable writes on `.kibana_n-1`
+4. (new) Delete the dummy "version lock" document from `.kibana_n-1`
+
+Since our documentation and server logs have implicitly encouraged users to
+rollback without using snapshots, many users might have to rely on these
+additional migration steps to perform a rollback. Since even the existing
+steps are error prone, introducing more steps will likely introduce more
+problems than what it solves.
+
+<details>
+  <summary><s>Minimize data loss with mixed 7.x versions</s></summary>
+  
+  1. All future versions of Kibana 7.x will use the `.kibana_saved_objects`
+   alias to locate the current index. If `.kibana_saved_objects` doesn't
+   exist, newer versions will fallback to reading `.kibana`.
```

Comment:
Am I correct in understanding that this fallback behavior only applicable to the migration window? If it happens also after migrations, we probably have a decent amount of code in Kibana that will break (anything that uses an ES client to access `.kibana` directly).

Reply:
Migrations will fallback to reading data out of `.kibana` during the migration window. But this does introduce a breaking change to how we interact with kibana's data, once started up all reads and writes should go to the concrete index to which `.kibana_saved_objects` pointed to at startup time. `.kibana` will point to an outdated read-only index so any writes here will fail. So we will have to be sure that all code that writes to `.kibana` today is also updated.

Reply:
> So we will have to be sure that all code that writes to `.kibana` today is also updated.

This sounds like potentially a lot of work. I hope I am wrong 😄 Maybe we can hack this within the `ElasticsearchService`?

Reply:
I didn't do a thorough audit, but I think the only places where we read directly from `.kibana` is when collecting some metrics. Plugins should use `kibana.index` config setting so we should be able to trace any usages.